phoenix-core/src/main/java/org/apache/phoenix/index/GlobalIndexChecker.java

```diff
@@ -0,0 +1,371 @@
+                    indexRowKey, 0, indexRowKey.length) == 0) {
+                if (!verifyRowAndRemoveVerifiedCell(row)) {
+                    // The corresponding row does not exist in the data table.
+                    // Need to delete the row from index if it is old enough
+                    deleteRowIfAgedEnough(indexRowKey, ts);
+                    row.clear();
+                }
+                return;
+            }
+            // This means the current index row is deleted by the rebuild process and we got the next row.
+            // If it is verified then we are good to go. If not, then we need to repair the new row
+            if (!verifyRowAndRemoveVerifiedCell(row)) {
+                // Rewind the scanner and let the row be scanned again so that it can be repaired
+                scanner.close();
+                scanner = region.getScanner(indexScan);
+                row.clear();
+            }
+        }
+
+        private boolean isVerifiedCell(Cell cell) {
+            return Bytes.compareTo(cell.getFamilyArray(), cell.getFamilyOffset(), cell.getFamilyLength(),
+                    emptyCF, 0, emptyCF.length) == 0 &&
+                    Bytes.compareTo(cell.getQualifierArray(), cell.getQualifierOffset(), cell.getQualifierLength(),
+                            emptyCQ, 0, emptyCQ.length) == 0;
+        }
+
+        private boolean verifyRow(byte[] rowKey) throws IOException {
+            Get get = new Get(rowKey);
+            get.setTimeRange(minTimestamp, maxTimestamp);
+            get.addColumn(emptyCF, emptyCQ);
+            Result result = region.get(get);
+            if (result.isEmpty()) {
+                LOG.debug("The empty column does not exist in a row in " + region.getRegionInfo().getTable().getNameAsString());
+                return true;
+            }
+            LOG.debug("Scan " + scan + " did not return the empty column for " + region.getRegionInfo().getTable().getNameAsString());
+            if (Bytes.compareTo(result.getValue(emptyCF, emptyCQ), 0, FALSE_BYTES.length,
+                    FALSE_BYTES, 0, FALSE_BYTES.length) == 0) {
+                return false;
+            }
+            return true;
+        }
+
+        private boolean verifyRowAndRemoveVerifiedCell(List<Cell> cellList) throws IOException {
+            long cellListSize = cellList.size();
+            Cell cell = null;
+            if (cellListSize == 0) {
+                return true;
+            }
+            Iterator<Cell> cellIterator = cellList.iterator();
+            while (cellIterator.hasNext()) {
+                cell = cellIterator.next();
+                if (isVerifiedCell(cell)) {
+                    if (Bytes.compareTo(cell.getValueArray(), cell.getValueOffset(), cell.getValueLength(),
+                            FALSE_BYTES, 0, FALSE_BYTES.length) == 0) {
+                        return false;
+                    }
+                    if (cellListSize > 1) {
```

Comment:
Please add comments as to why the (empty) column is getting removed.

Reply:
Will do